File: tools/modifythefilename.js
// modifythefilename.js

const { ipcRenderer } = require('electron');

let folderPath = '';                // 用來存儲文件路徑

export async function modifythefilenameFunction() {

    // 獲取 DOM 元素
    const contentDiv = document.getElementById('content');

    contentDiv.innerHTML = `

        <h1>自动修改文件名</h1>

        <div class="import">
            <div>
                <label style="width: 30%;">指定字符（不检查扩展名）</label>
                <input id="source_character" type="text">
            </div>

            <div>
                <label style="width: 30%;">指定字符的第一个字符的位置（从左至右，第几个字符）</label>
                <input id="location_character" type="text">
            </div>

            <div>
                <label style="width: 30%;">替换成（不替换扩展名）</label>
                <input id="target_character" type="text">
            </div>

            <div>
                <label style="width: 30%;">文件夹路径</label>
                <input id="folder_path" type="text">
            </div>

            <div style="display: flex; justify-content: center;">
                <button id="selectButton">选择文件夹</button>
                <button id="findButton">查找指定字符</button>
                <button id="addButton">添加指定字符</button>
                <button id="delButton">删除指定字符</button>
                <button id="replaceButton">替换指定字符</button>
                <button id="regexButton">使用正则表达式</button>
            </div>

            <div style="display: flex; flex-direction: column; align-items: center;">
                <label style="width: 100%;">&nbsp;&nbsp;说明：</label>
                <label style="width: 100%;">&nbsp;&nbsp;1.使用查找功能时，若指定字符的位置不填写，则检查文件名所有位置（不含扩展名）；</label>
                <label style="width: 100%;">&nbsp;&nbsp;2.使用添加功能时，若需要将指定字符插入原有文件名之前，则输入“0”；</label>
                <label style="width: 100%;">&nbsp;&nbsp;3.使用删除功能时，若指定字符的位置不填写，则对文件名中的所有位置（不含扩展名）进行删除；</label>
                <label style="width: 100%;">&nbsp;&nbsp;4.使用删除功能时，若填写指定字符的位置，则指定字符处需填写要删除的字符数；</label>
                <label style="width: 100%;">&nbsp;&nbsp;5.使用替换功能时，若位置不填写，则对文件名中的所有位置（不含扩展名）进行替换；</label>
                <label style="width: 100%;">&nbsp;&nbsp;6.使用正则表达式功能时，请使用Python语言的正则表达式规范。</label>
            </div>

        </div>

        <div class="export">
            <div>
                <label>查找/修改结果：</label>
            </div>
            <div>
                <textarea id="result_output" rows="10"></textarea>
            </div>
        </div>
        `;

    var input = document.getElementById('folder_path');
    input.classList.add('readonly');
    input.readOnly = true;

    var result_output = document.getElementById('result_output');
    result_output.classList.add('readonly');
    result_output.readOnly = true;

    document.getElementById('selectButton').addEventListener('click', async () => {
        // 向主進程發送請求，打開文件選擇對話框
        folderPath = await ipcRenderer.invoke('dialog:openDirectory');

        if (!folderPath) {
            console.log('Folder selection was canceled.');
            return;
        }

        document.getElementById(`folder_path`).value = folderPath;
    });

    const handleButtonClick = async (select_function) => {
        if (!folderPath) {
            alert('请先选择文件夹！');
            return;
        }

        const data = {
            select_function: select_function,
            source_character: document.getElementById('source_character').value,
            location_character: document.getElementById('location_character').value,
            target_character: document.getElementById('target_character').value,
            folderPath: folderPath
        };

        ipcRenderer.send('asynchronous-message', { command: 'filename_modify', data: data });
    };

    document.getElementById('findButton').addEventListener('click', () => handleButtonClick('find'));
    document.getElementById('addButton').addEventListener('click', () => handleButtonClick('add'));
    document.getElementById('delButton').addEventListener('click', () => handleButtonClick('del'));
    document.getElementById('replaceButton').addEventListener('click', () => handleButtonClick('replace'));
    document.getElementById('regexButton').addEventListener('click', () => handleButtonClick('regex'));

    ipcRenderer.removeAllListeners('asynchronous-reply');
    
    ipcRenderer.on('asynchronous-reply', (event, result) => {
        if (result[0] === 'filename_modify') {
            document.getElementById(`result_output`).value = '';
            document.getElementById(`result_output`).value = result[1]
        }
    });
}

File: tools/modifythefilename.py
# modifythefilename.py

import os
import re

def modify(request):
    select_function = request.get("data", {}).get("select_function", "")
    source_character = request.get("data", {}).get("source_character", "")
    location_character = request.get("data", {}).get("location_character", "")
    target_character = request.get("data", {}).get("target_character", "")
    folderPath = request.get("data", {}).get("folderPath", "")

    result_text = ''

    if select_function == 'find':

        if location_character != '':
            try:
                int(location_character)
            except:
                location_character = -1

        if location_character == '':
            for filename in os.listdir(folderPath):
                mainfilename = os.path.splitext(filename)[0]
                if source_character in mainfilename:
                    result_text = result_text + filename + '\n'
            if result_text == '':
                result_text = '未找到！'

        elif int(location_character) >= 1:
            location_character = int(location_character) - 1
            for filename in os.listdir(folderPath):
                if source_character == filename[location_character:location_character+len(source_character)]:
                    result_text = result_text + filename + '\n'
            if result_text == '':
                result_text = '未找到！'
        else:
            result_text = '参数输入有误，请重新输入！'

    elif select_function == 'add':

        try:
            int(location_character)
        except:
            location_character = -1

        if location_character == '':
            result_text = '参数输入有误，请重新输入！'

        elif int(location_character) == 0:
            for filename in os.listdir(folderPath):
                newfilename = source_character + filename
                oldfilePath = os.path.join(folderPath, filename)
                newfilePath = os.path.join(folderPath, newfilename)
                os.rename(oldfilePath, newfilePath)
                result_text = result_text + filename + ' to ' + newfilename + '\n'
            if result_text == '':
                result_text = '未找到可修改文件或文件夹！'

        elif int(location_character) >= 1:
            for filename in os.listdir(folderPath):
                if int(location_character) <= len(os.path.splitext(filename)[0]):
                    newfilename = filename[:int(location_character)-1] + source_character + filename[int(location_character)-1:]
                    oldfilePath = os.path.join(folderPath, filename)
                    newfilePath = os.path.join(folderPath, newfilename)
                    os.rename(oldfilePath, newfilePath)
                    result_text = result_text + filename + ' to ' + newfilename + '\n'
            if result_text == '':
                result_text = '未找到可修改文件或文件夹！'

        else:
            result_text = '参数输入有误，请重新输入！'

    elif select_function == 'del':

        if location_character != '':
            try:
                int(location_character)
            except:
                location_character = -1

        if location_character == '':
            for filename in os.listdir(folderPath):
                mainfilename = os.path.splitext(filename)[0]
                extfilename = os.path.splitext(filename)[1]
                if source_character in mainfilename:
                    newfilename = mainfilename.replace(source_character, '') + extfilename
                    oldfilePath = os.path.join(folderPath, filename)
                    newfilePath = os.path.join(folderPath, newfilename)
                    os.rename(oldfilePath, newfilePath)
                    result_text = result_text + filename + ' to ' + newfilename + '\n'
            if result_text == '':
                result_text = '未找到可修改文件或文件夹！'

        elif int(location_character) >= 1:
            try:
                int(source_character)
            except:
                source_character = -1

            if int(source_character) >= 1:
                for filename in os.listdir(folderPath):
                    mainfilename = os.path.splitext(filename)[0]
                    extfilename = os.path.splitext(filename)[1]
                    newfilename = filename[:int(location_character)-1] + filename[int(location_character)-1+int(source_character):]
                    oldfilePath = os.path.join(folderPath, filename)
                    newfilePath = os.path.join(folderPath, newfilename)
                    os.rename(oldfilePath, newfilePath)
                    result_text = result_text + filename + ' to ' + newfilename + '\n'
                if result_text == '':
                    result_text = '未找到可修改文件或文件夹！'
            
            else:
                result_text = '参数输入有误，请重新输入！'

    elif select_function == 'replace':

        if location_character != '':
            try:
                int(location_character)
            except:
                location_character = -1

        if location_character == '':
            for filename in os.listdir(folderPath):
                mainfilename = os.path.splitext(filename)[0]
                extfilename = os.path.splitext(filename)[1]
                if source_character in mainfilename:
                    newfilename = mainfilename.replace(source_character, target_character) + extfilename
                    oldfilePath = os.path.join(folderPath, filename)
                    newfilePath = os.path.join(folderPath, newfilename)
                    os.rename(oldfilePath, newfilePath)
                    result_text = result_text + filename + ' to ' + newfilename + '\n'
            if result_text == '':
                result_text = '未找到可修改文件或文件夹！'

        elif int(location_character) >= 1:
            for filename in os.listdir(folderPath):
                if source_character == filename[int(location_character)-1:int(location_character)-1+len(source_character)]:
                    newfilename = filename[:int(location_character)-1] + target_character + filename[int(location_character)-1+len(source_character):]
                    oldfilePath = os.path.join(folderPath, filename)
                    newfilePath = os.path.join(folderPath, newfilename)
                    os.rename(oldfilePath, newfilePath)
                    result_text = result_text + filename + ' to ' + newfilename + '\n'
            if result_text == '':
                result_text = '未找到可修改文件或文件夹！'
            
        else:
            result_text = '参数输入有误，请重新输入！'

    elif select_function == 'regex':

        try:
            if target_character == '':
                for filename in os.listdir(folderPath):
                    mainfilename = os.path.splitext(filename)[0]
                    regexcompile = re.compile(source_character)
                    if re.search(regexcompile,mainfilename):
                        result_text = result_text + filename + '\n'
                if result_text == '':
                    result_text = '未找到！'

            else:
                for filename in os.listdir(folderPath):
                    mainfilename = os.path.splitext(filename)[0]
                    extfilename = os.path.splitext(filename)[1]
                    regexcompile = re.compile(source_character)
                    if re.search(regexcompile, mainfilename):
                        newfilename = re.sub(regexcompile, target_character, mainfilename) + extfilename
                        oldfilePath = os.path.join(folderPath, filename)
                        newfilePath = os.path.join(folderPath, newfilename)
                        os.rename(oldfilePath, newfilePath)
                        result_text = result_text + filename + ' to ' + newfilename + '\n'
                if result_text == '':
                    result_text = '未找到可修改文件或文件夹！'
        except:
            result_text = '参数输入有误，请重新输入！'

    return ['filename_modify', result_text]